_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,605 @@
+
+{% highlight javascript %}
+// 3.2.6.b.1 AllTermsConstructively is a subset of any set satisfying the induction criteria (hence AllTermsConstructively <= AllTermsInductively)
+lemma AllTermsConstructivelyIncludedInSetsSatisfyingInductionCriteria(
+  terms: iset<Term>
+)
+  requires InductionCriteria(terms)
+  ensures AllTermsConstructively <= terms
+{
+  forall i: nat ensures S(i) <= terms {
+    InductionCriteriaHasConcreteIAsSubset(i, terms);
+  }
+}
+
+lemma AllTermsConstructivelyIncludedInAllTermsInductively()
+  ensures AllTermsConstructively <= AllTermsInductively
+{
+  InductiveAxioms();
+  AllTermsConstructivelyIncludedInSetsSatisfyingInductionCriteria(AllTermsInductively);
+}
+{% endhighlight %}
+
+## 4. Conclusion with the equality
+
+Because we have `<=` and `>=` between these two sets, we can now prove equality.
+
+{% highlight javascript %}
+lemma InductionAndConcreteAreTheSame()
+  ensures AllTermsConstructively == AllTermsInductively
+{
+  InductiveAxioms();
+  AllTermsConstructivelySatisfiesInductionCriteria();
+  AllTermsConstructivelyIncludedInAllTermsInductively();
+}
+{% endhighlight %}
+
+## Bonus Conclusion
+
+We were able to put together two definitions for infinite sets, and prove that these sets were equivalents.
+As stated in the introduction, having multiple definitions of a single infinite set makes it possible to pick the definition adequate to the job to prove other results. For example,
+
+- If a term is in the constructive set, then it cannot be `Add` for example, because it would need to be in a `S(i)` and none of the `S(i)` define `Add`. This can be illustrated in Dafny with the following lemma:
+
+{% highlight javascript %}
+lemma {:induction false} CannotBeAdd(t: Term)
+  requires t in AllTermsConstructively
+  ensures !t.Add?
+{
+  // InductiveAxioms(); // Uncomment if you use AllTermsInductively
+}
+{% endhighlight %}
+
+which Dafny can verify pretty easily. However, if you put `AllTermsInductively` instead of `AllTermsConstructively`, Dafny would have a hard time figuring out.
+
+- If `x` is in the inductive set, then `Succ(x)` is in the inductive set as well.
+Dafny can figure it out by itself using the `AllTermsInductively` definition, but won't be able to do it with `AllTermsConstructively` without a rigorous proof.
+
+{% highlight javascript %}
+lemma {:induction false} SuccIsInInductiveSet(t: Term)
+  requires t in AllTermsInductively
+  ensures Succ(t) in AllTermsInductively
+{
+  InductiveAxioms();
+}
+{% endhighlight %}
+
+This could be useful for a rewriter or an optimizer to ensure the elements it writes are in the same set.
+
+Everything said, everything above can be a bit overweight for regular Dafny users.
+In practice, you'd better off writing the inductive predicate explicitly as a function rather than an infinite set with a predicate, so that you get both inductive and constructive axioms that enable you to prove something similar to the two results above.
```

Comment:
```suggestion
In practice, you're better off writing the inductive predicate explicitly as a function rather than an infinite set with a predicate, so that you get both inductive and constructive axioms that enable you to prove something similar to the two results above.
```